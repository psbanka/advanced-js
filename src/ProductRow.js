import React, {Component} from 'react'

/**
 * converts category and name to a valid key string
 * @param {String} category - the name of the category
 * @param {String} name - the name of the string
 * @returns {String} - valid HTML id
 */
export function makeKey(category, name) {
  const convertedCategory = category.toLowerCase().replace(/ /g, '-')
  const convertedName = name.toLowerCase().replace(/ /g, '-')
  return `${convertedCategory}-${convertedName}`
}

export default class ProductRow extends Component {

  handleOnIsBuying (e) {
    const value = e.target.checked
    let key = `${this.props.currentCategory}${this.props.name}`
    this.props.onIsBuying(key, value, this.props.price)
  }

  render () {
    const filterMatch = this.props.name.indexOf(this.props.searchTerm) !== -1
    let style = {color: 'black'}
    if (!this.props.stocked) {
      style.color = 'red'
    }
    let key = makeKey(this.props.currentCategory, this.props.name)
    let amIChecked = this.props.isBuying[key] || false
    if (!this.props.inStock || this.props.stocked) {
      if (filterMatch) {
        return(
          <tr className="product" style={style}>
            <td>
              <input
                id={key}
                checked={amIChecked} type="checkbox"
                onChange={this.handleOnIsBuying.bind(this)}
             />{this.props.name}
            </td>
            <td>${this.props.price}</td>
          </tr>
        )
      }
    }
    return null
  }
}
