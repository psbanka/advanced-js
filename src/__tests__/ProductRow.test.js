import React from 'react'
import ReactDOM from 'react-dom'
import ProductRow, {makeKey} from '../ProductRow'
import { shallow, mount, render } from 'enzyme'
import renderer from 'react-test-renderer';

/* gloabl it describe */

describe('makeKey', () => {
  it('can make a key', () => {
    let output = makeKey('foo', 'bar')
    expect(output).toEqual('foo-bar')
  })
})

describe('ProductRow', () => {
  it('does a snapshot check', () => {
    const component = renderer.create(
      <ProductRow
        currentCategory='electronics'
        category='electronics'
        name='iphone 6'
        price={121}
        stocked={true}
        inStock={true}
        isBuying={{}}
        searchTerm=''
        onIsBuying={{}}
      />
    )
    const json = component.toJSON()
    expect(json).toMatchSnapshot()
  })

  describe('handleOnIsBuying', () => {
    it('properly calls upstream callback', () => {
      const callback = jest.fn()
      const wrapper = shallow(
        <ProductRow
          currentCategory='electronics'
          category='electronics'
          name='iphone 6'
          price={121}
          stocked={true}
          inStock={true}
          isBuying={{}}
          searchTerm=''
          onIsBuying={callback}
        />
      )
      const productRow = wrapper.instance()
      productRow.handleOnIsBuying({target: {checked: true}})
      expect(callback.mock.calls).toEqual([['electronicsiphone 6', true, 121]])
    })
  })

  describe('render', () => {
    it('performs basic callback functionality', () => {
      const callback = jest.fn()
      const wrapper = shallow(
        <ProductRow
          currentCategory='electronics'
          category='electronics'
          name='iphone 6'
          price={121}
          stocked={true}
          inStock={true}
          isBuying={{}}
          searchTerm=''
          onIsBuying={callback}
        />
      )
      const inputObject = wrapper.find('input')
      expect(inputObject.length).toBe(1)
      inputObject.simulate('change', {target: {checked: true}})
      expect(callback.mock.calls).toEqual([['electronicsiphone 6', true, 121]])
    })

    it('does not render if filtered out', () => {
      const wrapper = shallow(
        <ProductRow
          currentCategory='electronics'
          category='electronics'
          name='iphone 6'
          price={121}
          stocked={true}
          inStock={true}
          isBuying={{}}
          searchTerm='foo'
          onIsBuying={{}}
        />
      )
      const inputObject = wrapper.find('input')
      expect(inputObject.length).toBe(0)
    })

    it('does not render if not in stock and searching for stocked items', () => {
      const wrapper = shallow(
        <ProductRow
          currentCategory='electronics'
          category='electronics'
          name='iphone 6'
          price={121}
          stocked={false}
          inStock={true}
          isBuying={{}}
          searchTerm=''
          onIsBuying={{}}
        />
      )
      const inputObject = wrapper.find('input')
      expect(inputObject.length).toBe(0)
    })

    it('renders with red styling if out of stock', () => {
      const wrapper = shallow(
        <ProductRow
          currentCategory='electronics'
          category='electronics'
          name='iphone 6'
          price={121}
          stocked={false}
          inStock={false}
          isBuying={{}}
          searchTerm=''
          onIsBuying={{}}
        />
      )
      const inputObject = wrapper.find('.product')
      expect(inputObject.node.props.style.color).toBe('red')
    })

    it('renders with black styling if in stock', () => {
      const wrapper = shallow(
        <ProductRow
          currentCategory='electronics'
          category='electronics'
          name='iphone 6'
          price={121}
          stocked={true}
          inStock={false}
          isBuying={{}}
          searchTerm=''
          onIsBuying={{}}
        />
      )
      const inputObject = wrapper.find('.product')
      expect(inputObject.node.props.style.color).toBe('black')
    })
  })
})
