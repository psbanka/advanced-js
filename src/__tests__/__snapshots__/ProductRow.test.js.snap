// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductRow does a snapshot check 1`] = `
<tr
  className="product"
  style={
    Object {
      "color": "black",
    }
  }
>
  <td>
    <input
      checked={false}
      id="electronics-iphone-6"
      onChange={[Function]}
      type="checkbox"
    />
    iphone 6
  </td>
  <td>
    $
    121
  </td>
</tr>
`;

exports[`ProductRow snapshot check 1`] = `
<tr
  className="product"
  style={
    Object {
      "color": "black",
    }
  }
>
  <td>
    <input
      checked={false}
      id="electronics-iphone-6"
      onChange={[Function]}
      type="checkbox"
    />
    iphone 6
  </td>
  <td>
    $
    121
  </td>
</tr>
`;
